### PR TITLE
Add Rule Diagnostics

### PR DESCRIPTION
**Pull request by molly-moen**
# Type of change? 

- [ ] Bug fix
- [ ] New feature
- [x] Enhancement
- [ ] Breaking change (breaks backward compatibility)

# Description, Context, Motivation 
Add diagnostics logging for rule create/update events. Added the following metrics:
Rule_NewClick
Rule_EditClick
Rule_DeviceGroupClick
Rule_CalculationClick
Rule_FieldClick
Rule_OperatorClick
Rule_AddConditionClick
Rule_SeverityLevelClick
Rule_StatusToggle
Rule_ApplyClick
Rule_CancelClick
Rule_TopXCloseClick

Also includes new "sessionid" sections of diagnostics call, which logs the time in ms since Jan 1, 1970 when the page was loaded (amplitude expects session id in this format). This fields will be added by diagnostics to enable logging of session id to amplitude--until those changes go in it will be ignored by the backend. 

**Checklist:**

- [x] All tests passed
- [x] The code follows the code style and conventions of this project
- [ ] The change requires a change to the documentation
- [ ] I have updated the documentation accordingly

---
This change is [Reviewable](https://reviewable.io/reviews/azure/pcs-remote-monitoring-webui/1064)

### CODE REVIEW

File: src/store/reducers/appReducer.js [hunk 1]
```diff
@@ -141,7 +144,17 @@ export const epics = createEpicScenario({
       GitHubService.getReleaseInfo()
         .map(toActionCreator(redux.actions.getReleaseInformation, fromAction))
         .catch(handleError(fromAction))
+  },
+
+  updateCurrentWindow: {
```
Comment (isaac73) [inline src/store/reducers/appReducer.js]: change this into a reducer

File: src/store/reducers/appReducer.js [hunk 2]
```diff
@@ -43,10 +44,12 @@ export const epics = createEpicScenario({
     epic: ({ payload }, store) => {
       const diagnosticsOptIn = getDiagnosticsOptIn(store.getState());
       if (diagnosticsOptIn) {
+        payload.sessionId = getSessionId(store.getState());
+        payload.eventProperties.CurrentWindow = getCurrentWindow(store.getState());
         return DiagnosticsService.logEvent(payload)
         /* We don't want anymore action to be executed after this call
             and hence return empty observable */
-          .map(_ => Observable.empty())
+          .flatMap(_ => Observable.empty())
```
Comment (isaac73) [inline src/store/reducers/appReducer.js]: @spryor is this correct?

File: src/services/models/logEventModels.js [hunk 3]
```diff
@@ -2,11 +2,23 @@
...
+    ConditionCount: rule.conditions.length,
+    FirstFieldChosen: rule.conditions[0].field,
+    FirstOperatorChosen: rule.conditions[0].operator
+  };
+
+  return toDiagnosticsModel(eventName, metadata);
+}
+
+export const toSinglePropertyDiagnosticsModel = (eventName, propertyTitle, property) => {
+  var metadata = {};
```
Comment (isaac73) [inline src/services/models/logEventModels.js]: @spryor is there a shortcut for this two line assignment?

Comment (spryor) [reply src/services/models/logEventModels.js]:
```js
const metadata = { propertyTitle: property };
```

**Comment by molly-moen** [comment]
I think in order to keep the logging as flexible as possible for the future it would be best to not have generic logging in the form controls/state links. By keeping it within the editor it's very easy to add or remove extra metadata and be specific about what the rule and metadata names are. I agree that it would be nice to have an automated function, but I don't know a good way to do it without adding extra complexity and losing flexibility. I think it's also good to have all logging done in a consistent way.

**Pull request merged**